src/lib/device-state.ts

```diff
@@ -97,6 +97,30 @@ export const serviceInstallFromImage = (
 export const formatImageLocation = (imageLocation: string) =>
 	imageLocation.toLowerCase();
 
+export type DeviceConfigHook = (
+	configVars: Dictionary<string>,
+	osVersion: string,
+) => void;
+
+export interface HookRegistry {
+	deviceConfig: DeviceConfigHook[];
+}
+
+const STATE_ENDPOINT_HOOKS: HookRegistry = {
+	deviceConfig: [],
+};
+
+export function registerHook<
+	K extends keyof HookRegistry,
+	Fn extends typeof STATE_ENDPOINT_HOOKS[K][number]
+>(name: K, fn: Fn) {
+	STATE_ENDPOINT_HOOKS[name].push(fn);
+}
+
+function runHook(name: keyof HookRegistry, args: any[]) {
```

Comment:
You should be able to get strict types with
```suggestion
function runHook<T extends keyof HookRegistry>(name: T, args: Parameters<HookRegistry[T][number]>) {
```

Reply:
Ah nice, I searched for something like `Parameters` but didn't find it.